File: Go/day2/AOC.go
package main

import (
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

var Color_max = map[string]int{
	"blue":  14,
	"red":   12,
	"green": 13,
}

func checkGame(game string) bool {
	rounds := strings.Split(strings.Split(game, ":")[1], ";")
	for _, round := range rounds {
		colors := regexp.MustCompile("[0-9]+ (blue|red|green)").FindAllString(round, -1)
		for _, color := range colors {
			c := strings.Split(color, " ")
			num, _ := strconv.Atoi(c[0])
			if num > Color_max[c[1]] {
				return false
			}
		}
	}
	return true
}

func getFewest(game string) []int {
	rounds := strings.Split(strings.Split(game, ":")[1], ";")
	f_blue, f_red, f_green := 0, 0, 0
	for _, round := range rounds {
		colors := regexp.MustCompile("[0-9]+ (blue|red|green)").FindAllString(round, -1)
		for _, color := range colors {
			c := strings.Split(color, " ")
			num, _ := strconv.Atoi(c[0])
			if c[1] == "red" && num > f_red {
				f_red = num
			} else if c[1] == "blue" && num > f_blue {
				f_blue = num
			} else if c[1] == "green" && num > f_green {
				f_green = num
			}
		}
	}
	return []int{f_red, f_blue, f_green}
}

func part1(input []byte) int {
	correct := 0
	games := strings.Split(string(input), "\n")
	for i, game := range games {
		if checkGame(game) {
			correct += i + 1
		}

	}
	return correct
}

func part2(input []byte) int {
	power := 0
	games := strings.Split(string(input), "\n")
	for _, game := range games {
		vals := getFewest(game)
		power += vals[0] * vals[1] * vals[2]
	}
	return power
}

func main() {
	input, err := os.ReadFile("input.txt")
	if err != nil {
		panic(err)
	}
	part1_sum := part1(input)
	part2_sum := part2(input)
	fmt.Printf("Part 1: %d\n", part1_sum)
	fmt.Printf("Part 2: %d\n", part2_sum)
}

File: Python/day15/AOC.py
import re

def hash(input):
    curr = 0
    for ch in input:
        curr += ord(ch)
        curr = (curr * 17) % 256
    return curr

def part1(input):
    return sum(map(hash, input))

def part2(input):
    hash_map = {}
    for val in input:
        if "=" in val:
            lable, _ = val.split("=")
            index = hash(lable)
            if index not in hash_map:
                hash_map[index] = [val]
            else:
                for i in range(len(hash_map[index])):
                    if hash_map[index][i].split("=")[0] == lable:
                        hash_map[index][i] = val
                        break
                else:
                    hash_map[index].append(val)
        else:
            lable, _ = val.split("-")
            index = hash(lable)
            if index not in hash_map:
                continue
            else :
                for i in range(len(hash_map[index])):
                    if hash_map[index][i].split("=")[0] == lable:
                        hash_map[index].pop(i)
                        break
                if len(hash_map[index]) == 0:
                    hash_map.pop(index)
    
    total_power = 0
    for key in hash_map:
        for i, val in enumerate(hash_map[key]):
            _, val = val.split("=")
            power = (key + 1) * (i + 1) * int(val)
            total_power += power
            
    return total_power

def main():
    input = open("input.txt", "r").read().splitlines()[0].split(",")
    part1_sum = part1(input)
    part2_sum = part2(input)
    print(f"Part 1: {part1_sum}")
    print(f"Part 2: {part2_sum}")

if __name__ == "__main__":
    main()